File: py/subscriber.py
import sys
import time
import zenoh
from zenoh import Sample, Reliability

print("Opening session...")
session = zenoh.open(zenoh.Config())

key = "demo/example/**"

print("Declaring Subscriber on '{}'...".format(key))

def listener(sample: Sample):
    print(f">> [Subscriber] Received {sample.kind} ('{sample.key_expr}': '{sample.payload.decode('utf-8')}')")
    
# WARNING, you MUST store the return value in order for the subscription to work!!
# This is because if you don't, the reference counter will reach 0 and the subscription
# will be immediately undeclared.
sub = session.declare_subscriber(key, listener, reliability=Reliability.RELIABLE())

print("Enter 'q' to quit...")
c = '\0'
while c != 'q':
    c = sys.stdin.read(1)
    if c == '':
        time.sleep(1)

# Cleanup: note that even if you forget it, cleanup will happen automatically when 
# the reference counter reaches 0
sub.undeclare()
session.close()

File: rs/subscriber.rs
use std::time::Duration;
use zenoh::config::Config;
use zenoh::prelude::sync::SyncResolve;
use zenoh::prelude::KeyExpr;

fn main() {
    // Consult the full documentation at: https://docs.rs/zenoh/0.7.2-rc/zenoh/index.html

    println!("Opening session...");
    let session = zenoh::open(Config::default()).res_sync().unwrap();

    let key_expr = KeyExpr::try_from("demo/example/**").unwrap();

    println!("Declaring Subscriber on '{}'...", &key_expr);
    let subscriber = session.declare_subscriber(&key_expr).res_sync().unwrap();

    loop {
        if let Ok(sample) = subscriber.recv_timeout(Duration::from_secs(2)) {
            println!(
                ">> [Subscriber] Received {} ('{}': '{}')",
                sample.kind,
                sample.key_expr.as_str(),
                sample.value
            );
        } else {
            println!("Waiting for message!");
        }
    }
}
